### PR TITLE
pybridge: Convert PackagesChannel into a generator

### PR DESCRIPTION
**Pull request by martinpitt**
This achieves the same effect of making the sending asynchronous and chunked, without having to involve threads. This cleans up commit 97bd80971984.

### CODE REVIEW

**Comment by martinpitt** [comment]
Eww, this [wreaks havoc](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19145-20230728-074430-3581ab1c-fedora-38-devel/log.html) in the devel scenario with big package files. So apparently the GeneratorChannel isn't so async after all? This isn't urgent, the current threading solution works. If that doesn't work, then we should also reconsider the filesystems.py channels, though. Blocking there is less obvious as in our tests we don't read big files.

**Review by allisonkarlitskaya — COMMENTED** [review]
The issue here is that even though we now send as chunks and respect flow control, we're still not doing the file IO async.  I'm not sure if that's enough to cause the issue that the tests are running into, but maybe it is.

In the fsread channel (in addition to probably only using it on small files) we get a little bit of async behaviour by reading the file in chunks as regulated by flow control cutting us off and forcing us to wait for it to be received.  We're still blocking, but on smaller chunks, and (probably) letting the kernel prefetch some data for us in the background...

I'm going to try the approach discussed in standup: returning a file-like object (or maybe a file path including the possibility of `/dev/null`) in the `Document` object and doing (fake) async read via the asyncio executor.

File: src/cockpit/channels/packages.py [hunk 1]
```diff
@@ -108,14 +94,10 @@ def do_done(self) -> None:
...
-    def send_document_data(self, loop, data):
-        # split data into 4K blocks, to not overwhelm the channel
-        block_size = 4096
-        for i in range(0, len(data), block_size):
-            loop.call_soon_threadsafe(self.send_data, data[i:i + block_size])
-        loop.call_soon_threadsafe(self.done)
-        loop.call_soon_threadsafe(self.close)
+
+            # split data into 4K blocks, to not overwhelm the channel
+            block_size = 4096
```
Comment (allisonkarlitskaya) [inline src/cockpit/channels/packages.py:99]: In any case, this is the wrong choice — the flow control only activates at 16k boundaries.  You want `Channel.BLOCK_SIZE` here.

**Comment by martinpitt** [comment]
> The issue here is that even though we now send as chunks and respect flow control, we're still not doing the file IO async. I'm not sure if that's enough to cause the issue that the tests are running into, but maybe it is.

That can't be the sole reason. We've always synchronously read the full file in blocking mode, and it never looked [that bad](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19145-20230728-074430-3581ab1c-fedora-38-devel/log.html).

That said, making the reading async and moving it into the thread certainly should help. As discussed in the meeting, let's change `Document` from keeping a string to keeping a file object; then static strings can be wrapped into BytesIO.

**Comment by martinpitt** [comment]
@allisonkarlitskaya has something in the works, closing this.

**Pull request closed without merging**